tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -907,6 +908,98 @@ SpdpHandler::SpdpHandler(const Config& config,
 {}
 
+#ifdef OPENDDS_SECURITY
+namespace {
+  std::string extract_common_name(const OpenDDS::DCPS::Message_Block_Shared_Ptr& msg)
+  {
+    const CORBA::UShort encap_LE = 0x0300; // {PL_CDR_LE} in LE
+    const CORBA::UShort encap_BE = 0x0200; // {PL_CDR_BE} in LE
+
+    OpenDDS::RTPS::MessageParser message_parser(*msg);
+    message_parser.parseHeader();
+
+    while (message_parser.parseSubmessageHeader()) {
+      const auto submessage_header = message_parser.submessageHeader();
+      if (submessage_header.submessageId == OpenDDS::RTPS::DATA) {
+        unsigned short extraFlags;
+        unsigned short octetsToInlineQos;
+        OpenDDS::DCPS::EntityId_t readerId;
+        OpenDDS::DCPS::EntityId_t writerId;
+        OpenDDS::RTPS::SequenceNumber_t writerSequenceNumber;
+
+        if (!(message_parser >> extraFlags) ||
+            !(message_parser >> octetsToInlineQos) ||
+            !(message_parser >> readerId) ||
+            !(message_parser >> writerId) ||
+            !(message_parser >> writerSequenceNumber)) {
+          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: extract_common_name() could not parse submessage\n")));
+          return "";
+        }
+
+        if (writerId != OpenDDS::DCPS::ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER) {
+          // Not our message: this could be the same multicast group used
+          // for SEDP and other traffic.
+          break;
+        }
+
+        if (!message_parser.serializer().skip(octetsToInlineQos - 16)) {
+          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: extract_common_name() could not parse submessage\n")));
+          return "";
+        }
+
+        OpenDDS::RTPS::ParameterList inlineQos;
+        if ((submessage_header.flags & OpenDDS::RTPS::FLAG_Q) && !(message_parser >> inlineQos)) {
+          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: extract_common_name() could not parse submessage\n")));
+          return "";
+        }
+
+        if (submessage_header.flags & OpenDDS::RTPS::FLAG_D) {
+          OpenDDS::RTPS::ParameterList plist;
+
+          message_parser.serializer().swap_bytes(!ACE_CDR_BYTE_ORDER); // read "encap" itself in LE
+          CORBA::UShort encap, options;
+          if (!(message_parser >> encap) || (encap != encap_LE && encap != encap_BE)) {
+            ACE_ERROR((LM_ERROR,
+                       ACE_TEXT("(%P|%t) ERROR: extract_common_name() - ")
+                       ACE_TEXT("failed to deserialize encapsulation header for SPDP\n")));
+            return "";
+          }
+          message_parser >> options;
+          // bit 8 in encap is on if it's PL_CDR_LE
+          message_parser.serializer().swap_bytes(((encap & 0x100) >> 8) != ACE_CDR_BYTE_ORDER);
+          if (!(message_parser >> plist)) {
+            ACE_ERROR((LM_ERROR,
+                       ACE_TEXT("(%P|%t) ERROR: extract_common_name() - ")
+                       ACE_TEXT("failed to deserialize data payload for SPDP\n")));
+            return "";
+          }
```

Comment:
This can be replaced by deserializing an `EncapsualtionHeader` and getting the encoding from that.

Reply:
Let's handle that in a separate PR.